opm/simulators/gatherDeferredLogger.cpp

```diff
@@ -0,0 +1,146 @@
+
+namespace
+{
+
+    void packMessages(const std::vector<Opm::DeferredLogger::Message>& local_messages, std::vector<char>& buf, int& offset)
+    {
+
+        for (const auto lm : local_messages) {
+            MPI_Pack(&lm.flag, 1, MPI_INT64_T, buf.data(), buf.size(), &offset, MPI_COMM_WORLD);
+            int tagsize = lm.tag.size();
+            MPI_Pack(&tagsize, 1, MPI_UNSIGNED, buf.data(), buf.size(), &offset, MPI_COMM_WORLD);
+            MPI_Pack(lm.tag.c_str(), lm.tag.size(), MPI_CHAR, buf.data(), buf.size(), &offset, MPI_COMM_WORLD);
+            int textsize = lm.text.size();
+            MPI_Pack(&textsize, 1, MPI_UNSIGNED, buf.data(), buf.size(), &offset, MPI_COMM_WORLD);
+            MPI_Pack(lm.text.c_str(), lm.text.size(), MPI_CHAR, buf.data(), buf.size(), &offset, MPI_COMM_WORLD);
+        }
+    }
+
+    Opm::DeferredLogger::Message unpackSingleMessage(const std::vector<char>& recv_buffer, int& offset)
+    {
+        int64_t flag;
+        auto* data = const_cast<char*>(recv_buffer.data());
+        MPI_Unpack(data, recv_buffer.size(), &offset, &flag, 1, MPI_INT64_T, MPI_COMM_WORLD);
+
+        // unpack tag
+        unsigned int tag_length;
+        MPI_Unpack(data, recv_buffer.size(), &offset, &tag_length, 1, MPI_UNSIGNED, MPI_COMM_WORLD);
+        std::string tag;
+        if (tag_length>0) {
+            std::vector<char> tagchars(tag_length);
+            MPI_Unpack(data, recv_buffer.size(), &offset, tagchars.data(), tag_length, MPI_CHAR, MPI_COMM_WORLD);
+            tag = std::string(tagchars.data());
+        }
+        // unpack text
+        unsigned int text_length;
+        MPI_Unpack(data, recv_buffer.size(), &offset, &text_length, 1, MPI_UNSIGNED, MPI_COMM_WORLD);
+        std::string text;
+        if (text_length>0) {
+            std::vector<char> textchars(text_length);
+            MPI_Unpack(data, recv_buffer.size(), &offset, textchars.data(), text_length, MPI_CHAR, MPI_COMM_WORLD);
+            text = std::string (textchars.data());
+        }
+        return Opm::DeferredLogger::Message({flag, tag, text});
+    }
+
+    std::vector<Opm::DeferredLogger::Message> unpackMessages(const std::vector<char>& recv_buffer, const std::vector<int>& displ)
+    {
+        std::vector<Opm::DeferredLogger::Message> messages;
+        const int num_processes = displ.size() - 1;
+        for (int process = 0; process < num_processes; ++process) {
+            int offset = displ[process];
+            messages.push_back(unpackSingleMessage(recv_buffer, offset));
+            assert(offset == displ[process + 1]);
+        }
+        return messages;
+    }
+
+} // anonymous namespace
+
+
+namespace Opm
+{
+
+    /// combine (per-process) messages
+    Opm::DeferredLogger gatherDeferredLogger(const Opm::DeferredLogger& local_deferredlogger)
+    {
+        // Pack local messages.
+        int message_size = 0;
+        for (const auto lm : local_deferredlogger.messages_) {
+            message_size += sizeof(lm.flag);
```

Comment:
I guess one should use MPI_Pack_size here as the implementation might opt for padding

Reply:
done